clang/lib/CodeGen/SYCLLowerIR/LowerWGScope.cpp

```diff
@@ -0,0 +1,854 @@
+// For a kind 3 variable (alloca w/o metadata) this pass creates a WG-shared
+// local "shadow" variable. Before each PFWI invocation leader WI stores its
+// private copy of the variable into the shadow (under "is leader" guard), then
+// all WIs (ouside of "is leader" guard) load the shadow value into their
+// private copies ("materialize" the private copy). This works becase these
+// variables are uniform - i.e. have the same value in all WIs and are not
+// changed within PFWI. The only exceptions are captures of private_memory
+// isntances - see next.
+// ** Kind 1:
+// Even though WG-scope locals are supposed to be uniform, there is one
+// exception - capture of local of kind 1. It is always captured by non-const
+// reference because as there no
+// 'const T &operator()(const h_item<Dimensions> &id);' which means the result
+// of kind 1 variable's alloca is stored within the PFWI lambda.
+// Materialization of the lambda object value writes result of alloca of the
+// leader WI's private variable into the private copy of the lambda object,
+// which is wrong. So for tese variables this pass adds a write of the private
+// variable's address into the private copy of the lambda object right after its
+// materialization:
+//     if (is_leader())
+//       *PFWI_lambda_obj_shadow_addr = *PFWI_lambda_obj_alloca;
+//     barrier();
+// (1) *PFWI_lambda_obj_alloca = *PFWI_lambda_obj_shadow_addr;
+// (2) PFWI_lambda_obj_alloca->priv_var_addr = priv_var_alloca;
+//     parallel_for_work_item(..., PFWI_lambda_obj_alloca);
+//
+// (1) - materialization of a PFWI object
+// (2) - "fixup" of the private variable address.
+//
+// TODO The approach employed by this pass generates lots of barriers and data
+// copying between private and local memory, which might not be efficient. There
+// are optimization opportunities listed below. Also other aproaches can be
+// considered like
+// "Efficient Fork-Join on GPUs through Warp Specialization" by Arpith C. Jacob
+// et. al.
+//===----------------------------------------------------------------------===//
+
+#include "LowerWGScope.h"
+
+#include "clang/Basic/AddressSpaces.h"
+
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/Statistic.h"
+#include "llvm/ADT/Triple.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+
+#ifndef NDEBUG
+#include "llvm/IR/CFG.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/GraphWriter.h"
+#endif
+
+using namespace llvm;
+
+#define DEBUG_TYPE "lowerwgcode"
+
+STATISTIC(LocalMemUsed, "amount of additional local memory used for sharing");
+
+const char *WG_SCOPE_MD = "work_group_scope";
+const char *WI_SCOPE_MD = "work_item_scope";
+const char *PFWI_MD = "parallel_for_work_item";
+
+namespace {
+cl::opt<int> Debug("sycl-lower-wg-debug", llvm::cl::Optional, llvm::cl::Hidden,
```

Comment:
Please change this to static as well. http://llvm.org/docs/CodingStandards.html#anonymous-namespaces

Reply:
ok